### DISCUSSION TITLE
Read VCC supply level without external components

### DISCUSSION THREAD

**Discussion by LaZsolt** [dbuezas/lgt8fx#123] [Show and tell]
I modified the **Arduino_Vcc** library and now could work on LGT8F328P microcontroller: https://github.com/LaZsolt/Arduino_Vcc
Now you can measure your AVR microcontroller's Vcc even if it is a LogicGreen.

The story:
I bought a new Arduino compatible nano style board with **ATmega MCU**. I started to use it, but it regularly stopped. It was powered by USB charger. One day later I discovered, the power source selector diode dropped the USB voltage by 0.5 V so the Vcc was only 4.5 V. This Vcc not enough for a safe operation on 16 MHz. At the end I applied a hihger voltage power supply on the VIN pin.

So good to know your board's Vcc.

- But this modified **Arduino_Vcc** library couldn't work on LGT8F328D version microcontroller.

Edit:
However if you using LGT8F328P or D microcontrollers, reading the Vcc supply level is much easier, but must use board package v2.0.0 or newer. Watch this:
```
// filename: Vcctest2.ino
void setup() {

  Serial.begin(38400);

  // using internal 1.024 V or 1.25 V reference
#if defined(__LGT8FX8P__)
  analogReference(INTERNAL1V024);
#else
  analogReference(INTERNAL1V25);
#endif
  analogReadResolution(12);
} // end setup

void loop() {

  uint32_t v = analogRead(VCCM);
  
  // convert to VCC (unit: mV);
#if defined(__LGT8FX8P__)
  v = (v * 5 ) / 4 ;
#else
  v = (v * 5000 ) / 4096 ;
#endif

  // The maximum 
  Serial.print("Analog Power Supply: ");
  Serial.print(v, DEC);
  Serial.println(" mV");

  delay(500);
} // end loop
```

The `lgt8f328p_qfp48_case02.ino` sketch gave me the idea, from libraries of this lgt8fx package.

**Comment by SuperUserNameMan** [comment]
Very cool feature ! 
I did not know it was possible.
Going to integrate that into my projects now :-D

**Comment by jayzakk** [comment]
@LaZsolt , should be added to the compatible library list in the wiki, if it's ok for you?

**Reply by LaZsolt** [reply]
Yes, it's okay. I will happy about it.

**Reply by youxiaojie** [reply]
BUGS!! in write_analogue.c

```
void analogReference(uint8_t mode)
{
	// can't actually set the register here because the default setting
	// will connect AVCC and the AREF pin, which would cause a short if
	// there's something connected to AREF.
	analog_reference = mode;

#if defined(__LGT8FX8E__) || defined(__LGT8FX8P__)
	#if defined(__LGT8FX8E__)
	if(analog_reference == INTERNAL2V56) {
		VCAL = VCAL2;
	} else {
		VCAL = VCAL1;
	}
	#else
	// set analog reference for ADC/DAC
	if(analog_reference == EXTERNAL) {
		DACON = (DACON & 0x0C) | 0x1;
		if((PMX2 & 0x2) == 0x2) {
			GPIOR0 = PMX2 & 0xfd;
			PMX2 = 0x80; 
			PMX2 = GPIOR0;
		}
	} else if(analog_reference == DEFAULT) {
		DACON &= 0x0C;
	} else {
		DACON = (DACON & 0x0C) | 0x2;
		cbi(ADCSRD, REFS2);
		if(analog_reference == INTERNAL2V048) {
			VCAL = VCAL2;	// 2.048V
		} else if(analog_reference == INTERNAL4V096) {
			sbi(ADCSRD, REFS2);
			VCAL = VCAL3;	// 4.096V
		} else	{
			VCAL = VCAL1;	// 1.024V
		}
	}
	#endif

	ADMUX = (analog_reference << 6);
#endif
}


```
dear friend,
I looked at this code,find that if I was 328d, there is a sentence on page 204, 
使用内部基准(1.25V/2.56V)注意事项:
芯片上电后,默认将内部基准校准为 1.25V,用户如果使用 1.25V 的内部基准,可以直
接使用,无需其他操作。但如果需要使用 2.56V 的内部参考电压,需要自行更新内部基准的
校准值。 2.56V 的校准值在上电后被加载到寄存器 VCAL2(0xCE),在程序初始化时,将 VCAL2
的值读入并写入到 VCAL(0XC8)寄存器即完成 2.56V 的校准。
需要注意的是,当更新了 VCAL 寄存器后,会导致 VCAL1 即 1.25V 的校准值同时更新为
VCAL2,因此如果需要在后续使用过程中重新选择 1.25V 基准,需要提前将 VCAL1 的值保存
到变量中以供后续使用。

after power on, system set calibrate to 1v25 as default, no other operation needed. to using 2V56 as VCAL, need read calibrate value from VCAL2(0xce) and write to VCAL(0xC8). However when update the calibrate of VCL, the calibrate value in VCL1 will update to VCL2 simultaneously , so we must save calibrate of VCL1 in case we use 1v25 again after changing to 2V56.

```
so there seems a bug:
	#if defined(__LGT8FX8E__)
	if(analog_reference == INTERNAL2V56) {
		VCAL = VCAL2;
	} else {
		VCAL = VCAL1; // if  we use 1v25 again after changing to 2V56. VCAL1 store the calibrate value of 2V56
	}
	#else
```
To workout, I see in the main.cpp,GPIOR1 save the calibration,
 
```
main.cpp-#if defined(__LGT8FX8E__)
main.cpp-// store ivref calibration 
main.cpp:       GPIOR1 = VCAL1;
main.cpp:       GPIOR2 = VCAL2;
main.cpp-

```
can we replace VCAL1 with GPIOR1 here to workout this problem?
```

	#if defined(__LGT8FX8E__)
	if(analog_reference == INTERNAL2V56) {
		VCAL = VCAL2;
	} else {
		VCAL = GPIOR1; // will it get some side effect? 
	}if 
	#else
```

or are there some other good way to work around? does GPIOR1 be overwritten by other program?
Thanks!

**Reply by LaZsolt** [reply]
> can we replace VCAL1 with GPIOR1 here to workout this problem?

No, we can't replace VCAL1 with GPIORx because GPIORx registers are regularly overwritten. These registers are regularly used for temporary data storage.

**Reply by LaZsolt** [reply]
> No, we can't replace VCAL1 with GPIORx because GPIORx registers are regularly overwritten. These registers are regularly used for temporary data storage.

I checked the code and seems GPIOR1 and GPIOR2 not owerwritten. But not safe.

**Reply by youxiaojie** [reply]
do you have idea to solve it? to use a safer way.If the user use gpior1, it is not safe.

**Reply by LaZsolt** [reply]
I figured out the safer way is, to save VCAL1 content into a dedicated variable. I named it VCAL_1
This variable defined in `Arduino.h`, after reset, VCAL1 saved in `main.cpp` and used in `wiring_analog.c`

I will do a pull request in a few days but in the meantime here is the changes I made.
`Arduino.h`
```
#include "pins_arduino.h"

// For store ivref calibration
#if defined(__LGT8FX8E__)
extern uint8_t VCAL_1;
#define VCAL_2 VCAL2
#endif

#if defined(__LGT8FX8E__) || defined(__LGT8FX8P__)
#include "fastio_digital.h"
```

`main.cpp`
```
void lgt8fx8x_init()
{
#if defined(__LGT8F_SSOP20__)
	GPIOR0 = PMXCR | 0x07;
	PMXCR = 0x80;
	PMXCR = GPIOR0;
#endif

#if defined(__LGT8FX8E__)
// store ivref calibration value (According to the LGT8F328D databook, the value of VCAL1 is lost, if VCAL is modified/rewritten.)
	VCAL_1 = VCAL1;
//  VCAL_2 is defined in Arduino.h

// enable 1KB E2PROM 
	ECCR = 0x80;
	ECCR = 0x40;

// clock source settings
```

`wiring_analog.c`
```
void analogReference(uint8_t mode)
{
	// can't actually set the register here because the default setting
	// will connect AVCC and the AREF pin, which would cause a short if
	// there's something connected to AREF. (ATmega MCUs)
	analog_reference = mode;

#if defined(__LGT8FX8E__) || defined(__LGT8FX8P__)
	#if defined(__LGT8FX8E__)
	if(analog_reference == INTERNAL2V56) {
		VCAL = VCAL_2;
	} else {
		VCAL = VCAL_1;
	}
	#else
	// set analog reference for ADC/DAC
```

**Comment by youxiaojie** [comment]
I run in my 328d, show 1.48v? I will meaure it by multimeter to confirm.

**Reply by LaZsolt** [reply]
1.48V does not seems correct.
I haven’t tested the Vcc reader on 328D, because not received yet. But I ordered one.

**Reply by youxiaojie** [reply]
not correct. 4V7 by multimeter. so shows not correct.
by the way,massduino is produced by lgt. lgt's taobao shop also sell md* chips.

**Reply by youxiaojie** [reply]
I decide to order some 328d and 328p 32pin and 48pin and draw pcb,making my arduino. if you have interesting, I can share to you. the chip is very 3RMB for 32pin 328p 4RMB for 48pin.

**Comment by LaZsolt** [comment]
@youxiaojie 

Okay, I found some errors in `Vcc.cpp` and I corrected it.
So if you want to test, dowload `Vcc.cpp` again.

**Reply by youxiaojie** [reply]
VCC = inf Volts
VCC = 100.00 %
VCC = inf Volts

**Reply by LaZsolt** [reply]
Still not a good result. I am thinking what could be wrong....

**Reply by LaZsolt** [reply]
I have an idea. Try this:
- Overwrite your `C:\Users\_your_username_\AppData\Local\Arduino15\packages\LGT8fx Boards\hardware\avr\1.0.6\cores\lgt8f\wiring_analog.c` file with this new version https://github.com/dbuezas/lgt8fx/blob/815cc9f890fba9334ddfc91a09483a34e93e0037/lgt8f/cores/lgt8f/wiring_analog.c
- After that run the program what I inserted on the top of this page :

**Reply by youxiaojie** [reply]
Arduino: 1.8.13 (Linux), Board: "LGT8F328, 19200, Default (64), Internal, 16 MHz, 328D (rare)"

"/usr/bin/avr-gcc" -c -g -Os  -std=gnu11 -ffunction-sections -fdata-sections -MMD -flto -fno-fat-lto-objects -mmcu=atmega328p -DSERIAL_RX_BUFFER_SIZE=64 -DCLOCK_SOURCE=1 -DF_CPU=16000000L -DARDUINO=10813 -DARDUINO_AVR_LARDU_328E -DARDUINO_ARCH_AVR   "-I/home/allan/.arduino15/packages/LGT8fx Boards/hardware/avr/1.0.6/cores/lgt8f" "-I/home/allan/.arduino15/packages/LGT8fx Boards/hardware/avr/1.0.6/variants/lgt8fx8e" "/home/allan/.arduino15/packages/LGT8fx Boards/hardware/avr/1.0.6/cores/lgt8f/wiring_analog.c" -o "/tmp/arduino_build_815803/core/wiring_analog.c.o"
/home/allan/.arduino15/packages/LGT8fx Boards/hardware/avr/1.0.6/cores/lgt8f/wiring_analog.c: In function ‘__analogRead’:
/home/allan/.arduino15/packages/LGT8fx Boards/hardware/avr/1.0.6/cores/lgt8f/wiring_analog.c:149:33: error: expected expression before ‘||’ token
  if (( pin >= 9 && pin <= 12 )) || ( pin >= 16 )) return 0;   // invalid ADMUX selection in LGT8F328D
                                 ^
/home/allan/.arduino15/packages/LGT8fx Boards/hardware/avr/1.0.6/cores/lgt8f/wiring_analog.c:149:49: error: expected statement before ‘)’ token
  if (( pin >= 9 && pin <= 12 )) || ( pin >= 16 )) return 0;   // invalid ADMUX selection in LGT8F328D
                                                 ^
Using library Arduino_Vcc-master in folder: /home/allan/Arduino/libraries/Arduino_Vcc-master (legacy)
exit status 1
Error compiling for board LGT8F328.

after delete a ")" compile and run:

Analog Power Supply: = 4537mV
Analog Power Supply: = 4532mV
Analog Power Supply: = 4530mV

**Reply by LaZsolt** [reply]
Thank you for your test.
```
if (( pin >= 9 && pin <= 12 )) || ( pin >= 16 )) return 0; // invalid ADMUX selection in LGT8F328D
                             ^
                      Delete this round bracket.
```

**Reply by youxiaojie** [reply]
sure, this small case I can deal with:) thanks for your effort! 
to use the new code! I test it on 328p again.

**Reply by LaZsolt** [reply]
> after delete a ")" compile and run:
> 
> Analog Power Supply: = 4537mV

I am happy about the good result.

**Reply by youxiaojie** [reply]
seems higher than what multi-meter gets. about 4.34 actually. it does not matter. the diode take about 0.6V, I dont know why need this diode. if in case of inverted + and -,they can take a 5V TVS diode inverted parallel connected in + and -.(nagetive polar TVS connect with power + ) when the power connected wrongly, the current will pass through TVS diode, the voltage nearly zero, to protect following component. I will drop it! connect by a wire! mod myself.

**Reply by youxiaojie** [reply]
is the voltage effect frequency of core?
this is 328p result

VCC = 4.68 Volts
VCC = 93.52 %
Analog Power Supply: = 4667mV
Analog Power Supply: = 4673mV
Analog Power Supply: = 4676mV

**Reply by LaZsolt** [reply]
> seems higher than what multi-meter gets. about 4.34 actually.

I think I know the reason. According my measures, 328D has no such a big gain error then calculated in `wiring_analog.c` line 173.
If you delete this line then measured VCC in mV will be better.

**Reply by youxiaojie** [reply]
I will get an accurate meter to measure it again, maybe it is my fault. dont mind it.

**Comment by G4ngst4** [comment]
Hi, what is this VCCM pin?
I couldn't find any info on  this 
Thank you very much

**Reply by LaZsolt** [reply]
VCCM is a name of a virtual Arduino pin. This is an internal ADC MUX input for measuring AVCC level.
| LGT8F328D ADC | LGT8F328P ADC |
| - | - |
| ![328D-VCCM](https://user-images.githubusercontent.com/60486174/115426994-55052680-a201-11eb-9ba5-0e0fd2b6bef1.png) | ![328P-VCCM](https://user-images.githubusercontent.com/60486174/115427025-5d5d6180-a201-11eb-9503-90ced6a7d9b9.png) |

click on the image to enlarge 

Edit: VCCM is a name of a virtual Arduino pin.

**Reply by LaZsolt** [reply]
As I mentioned above must use newer board package than v1.0.6 which is not released yet. (april 20, 2021)
I can help if you want the necessary files.

**Reply by G4ngst4** [reply]
Many thanks for your help!
Yes I am trying to understand this and I can't find any info or datasheet with pictures like you just gave me, 
Yes, I'd love to have the updated files and any technical docs that can help, many thanks for your help!

**Reply by LaZsolt** [reply]
Replace your `C:\Users\_your_username_\AppData\Local\Arduino15\packages\LGT8fx Boards\hardware\avr\1.0.6\cores\lgt8f\wiring_analog.c` file with this new version
https://github.com/dbuezas/lgt8fx/blob/815cc9f890fba9334ddfc91a09483a34e93e0037/lgt8f/cores/lgt8f/wiring_analog.c

**Reply by G4ngst4** [reply]
Thank you , I will try it and try understand it. 
Looking at the diagram and code if I managed to understand it correctly, this reads some 1/5 of avcc pin (this has a built-in voltage divider??) against the internal 1.024 reference? 
I'm not sure I understand as new to this & trying to find technical information on this LGT8F328P chip & boards.

**Reply by LaZsolt** [reply]
> This has a built-in voltage divider??

Yes.

Technical informations:

![kép](https://user-images.githubusercontent.com/60486174/115579333-59474780-a2c6-11eb-961e-220c8a408ac3.png)

**Reply by G4ngst4** [reply]
I see. Wondering if using this can enable measurement of higher voltages than vcc without damage, you think it's possible 🤔?

**Reply by LaZsolt** [reply]
![kép](https://user-images.githubusercontent.com/60486174/115610365-0df26080-a2e9-11eb-80a0-c257fdeb60b3.png)

**Reply by G4ngst4** [reply]
Thanks, got too hopeful seeing this I guess 😄

**Reply by jayzakk** [reply]
Correct, VCC is the limit due to the input select matrix. If you need to measure higher voltage, you need to add a resistor divider in front of the input.

**Reply by dbuezas** [reply]
Note: there are internal voltage divisors, but even if you use them, the max voltage that can be read is still VCC for whatever reason. I tried it in the oscilloscope project [here](https://github.com/dbuezas/arduino-web-oscilloscope/blob/master/arduino/src/adc.h#L27-L28)

**Reply by G4ngst4** [reply]
@dbuezas looks amazing. way over my head figuring this out though, just got the board and new to this.
So just for clarification, when you tried measuring higher voltage, you used the internal divisors without an external voltage divider resistors? Against a vcc reference?
And figured out the voltage from there? 
That failed to work? 
You have some basic example /link showing this?
Btw MANY thanks for this library!

**Reply by rokweom** [reply]
> Correct, VCC is the limit due to the input select matrix. If you need to measure higher voltage, you need to add a resistor divider in front of the input.

Sorry for the offtopic, but does that mean that I can measure voltages up to VCC (5V) even when using 4.096V internal reference? The problem I'm having right now is that I need very accurate voltage measurements, but I'm powering the board from USB so I can't be certain that the VCC is exactly 5V.

**Reply by rokweom** [reply]
To answer myself: no, you can't. Which obviously makes perfect sense, but @jayzakk 's comment got me a little confused.

**Reply by LaZsolt** [reply]
> Can I measure voltages up to VCC (5V) even when using 4.096V internal reference?

Yes the LGT8F328P can. The example program presented at the beginning of this topic does just that. The trick is the the using of the LGT8F328 capability: the ADC has an 1/5 of VCC input. This internal input pin name: VCCM
With this method the MCU can measure its own VCC up to 5.1V
The example program uses 1.024V as reference because it will work better if accidently the VCC drops below 4.1V
But if you want to use 4.096V as a reference then must be use the ADC's 4/5 of VCC internal input, which name is V5D4.
`uint32_t v = analogRead(V5D4);`

**Reply by LaZsolt** [reply]
Oops. My answer is not appropriate to your question.

**Reply by rokweom** [reply]
Yup :) But it's fine, because once you measure the VCC using the arduino_vcc library, you can then use that as a reference and still get accurate readings of any voltage up to VCC.

**Comment by ADI84211** [comment]
Hi Guys,
Hope I'm not intruding here.
I have recently started programming the LGT8F328P in  assembly language using Atmel (Microchip) Studio 7 and programming my hex files with AVRdudess. I sorted out the set up for 32mhz RC oscillator and set up for divide by 1 prescaler.
I have the DAC working fine except for the 2.048v and 4.096v references.
I've had no problems with VCC  or VCAL1 1.024v  as Vref but the other 2 just dont want to play ball.
Is there something I'm missing ? Reading the not so clear English translations of the Chinese data sheet, I figured all I needed to do was write the value in VCAL2 or VCAL3 to VCAL register after power up. I've given the micro plenty of time for voltages to settle after 
 power up, before  writing to VCAL.  Any help, much appreciated. Cheers Jim

**Comment by LaZsolt** [comment]
Hi Jim!

> Reading the not so clear English translations of the Chinese data sheet

Yes, the datasheet is'nt clear. As I figured out, the internal voltage source has 3 outputs (1.024V, 2.048V, 4.096V), but only has one common calibrating register. So when you write the VCAL regiser, it doesn't select the reference output voltage, it just adjusting it.

> I figured all I needed to do was write the value in VCAL2 or VCAL3 to VCAL register after power up.

You can write the VCAL register any time, not only after power up, but mostly before you start using the internal voltage source.

Along with setting the VCAL the IVSEL0 and IVSEL1 bits of the ADCSRD register must also be set. (Fifth and fourth bits.) But ADC reference voltage is must be set as AVCC or AVREF. (ADMUX register)
**See the table below.**

> I have the DAC working fine except for the 2.048v and 4.096v references.

Before you start using DAC, it's reference voltage source muts be selected.  Set DACON register bit DAVS1 to 1 and DAVS0 to 0.

![kép](https://user-images.githubusercontent.com/60486174/210236160-bec8c30f-a6f9-4381-b314-1e3eb7841732.png)

(I didn't know these either until now.)

**Reply by LaZsolt** [reply]
On the other hand, this is not a VCC level measurement topic either.

**Reply by ADI84211** [reply]
Hello LaZsoft,
Thank you for your valuable info. I'll try again today armed with this info. 
Best regards from New Zealand.
Cheers Jim

**Reply by ADI84211** [reply]
Hello LaZsoft,
Got it working correctly, many thanks to your help.
Best regards Jim

**Comment by whyameye** [comment]
Slightly off topic but I noticed this statement: 

>  Vcc was only 4.5 V. This Vcc not enough for a safe operation on 16 MHz.

It sounds like you are saying LGT8Fx cannot operate reliably when Vcc for is 4.5V or less and it is operating at 16Mhz. I couldn't find verification of this in the datasheet. Is it there or are you basing this on your emperical evidence? If <4.5V is not enough, what is the correct voltage for 16Mhz? I assume the voltage need to be the same or higher for 32Mhz? Am I correct that I cannot run below 32Mhz without adding an external oscillator?

Context: currently I have a circuit where the LGT8Fx is powered at 3.3V and running on the internal 32Mhz  and I'm wondering if some of the flakey behavior I'm seeing could be related to too low a voltage supply. It would not be trivial to test because of other components but I'm up for it if this could be the problem.

**Reply by dbuezas** [reply]
If I recall correctly, the lgt328p can run at full 32Mhz at 1.8v. He probably referred to the atmega. Although the atmega328p should be able to run at 20Mhz at 4.5v 🤔

**Reply by LaZsolt** [reply]
> I bought a new Arduino compatible nano style board with ATmega MCU.  ...  This Vcc not enough for a safe operation on 16 MHz.

On this board was an ATmega 168 MCU. The maximum operating frequency is lower than ATmega 328P. At the end the power applied not by USB.

![kép](https://github.com/dbuezas/lgt8fx/assets/60486174/5552a1d1-ba45-4714-9f2e-b8e435a2dfb3)

**Comment by Portia-Lin** [comment]
Hello. I would like to ask if it is possible to switch between the internal reference voltage source and the VCC voltage (external) when using the analog input to read the values without damaging the microcontroller. I'm using a project where the analog input goes from 0V to almost the supply voltage. I don't use voltage dividers. Can I stop reading the analog input, switch to an internal voltage reference, measure the supply voltage with your code, switch to an external voltage reference, and read the analog input again? The voltage at the analog input is greater than the voltage of the internal reference voltage source. I can't use the internal 4.096V reference voltage source because the microcontroller is battery powered. Will my plan damage the microcontroller?

**Reply by dbuezas** [reply]
It shouldn't damage anything, just make sure you don't measure above supply.
To change the analog reference you can call e.g:
```cpp
analogReference(DEFAULT);  // supply
  // analogReference(EXTERNAL);      // REF PIN Voltage
  // analogReference(INTERNAL4V096); // 4.096V
  // analogReference(INTERNAL2V048); // 2.048v
  // analogReference(INTERNAL1V048); // 1.024v
```

Also, if you are powering via usb, remember there is a diode in between so your supply voltage will be around 0.6v velow 5v

**Reply by Portia-Lin** [reply]
I am developing a device that is powered by a battery with a maximum voltage of 4.2V. I plan to know the battery voltage using your code, at this time I switch to the internal reference voltage (1.024V). After that, I switch to the reference voltage of the power supply and measure the ADC value. At this time there is more than 1.024V coming to the analog input, but I am not taking any readings from the input. And it's safe, am I right? I ask this so much because I once burned one analog input on an atmega328 playing with these things. I also attached a piece of code, if you can please reply if it works or not. For example, the voltage at the A0 input is constantly within 2-3V, and the reference voltage source is periodically switched (1.024V / DEFAULT). Thank you.

```
uint32_t v = 0; // Supply voltage in mV
int adc_v = 0;  // ADC value at input A0

void setup() {
  // put your setup code here, to run once:
  Serial.begin(38400);
  analogReadResolution(12);

}

void loop() {
  // Measuring the supply voltage
  analogReference(INTERNAL1V024);
  delay(5);
  v = analogRead(VCCM);
  v = (v * 5 ) / 4;
  Serial.print("Analog Power Supply: ");
  Serial.print(v, DEC);
  Serial.println(" mV");
  analogReference(DEFAULT);
  delay(5);
  // Measuring the adc value
  adc_v = analogRead(A0);
  Serial.print("ADC value: ");
  Serial.println(adc_v);
  delay(1000);
}
```

**Comment by gadefox** [comment]
VCCM=22; V5D1=22 so `analogRead` will set `ADCSRD` to `0b?????110` (Vcc) and `ADMUX` to `0b11?01000` (1/5VDO and 1.024V reference)/`0b10?01000` (1/5VDO and 2.048V reference). ADC=3787 (4.734V)/1891 (4.728V) which is fine due to a diode (LGT8F nano module).

1) I'm getting ADC=0 in case of 4.096V reference.
2) The same with 4/5VDO.

Any idea??

EDIT:
1) It works fine.. The output is 943 (4.715V)
```
ADCSRD = 0b?1???110 // REFS2 + VDS: 110 (system power)
ADMUX = 0b00?01000; // 4.096V ref. + 1/5VDO
VCAL = VCAL3;
```

2) Works only with 4.096V reference. `Vcc=ADC*1.25 [mV]` (4.730V)
